File: Practice/atcoder/ABC/070/src/d.py
# disjoint setかと思いきやDFS
# したらTLEになったので根本的にやり方が間違ってるということ
# よくよく考えたらdfsなのでそこにたどり着いたときが最短距離だわ ムダに探索してた
# Pypyだと残り1TLE, Pythonだとちょっときついけど間に合う感じになった。。

import sys

sys.setrecursionlimit(10000000)

n = int(input())
graph = [[] for _ in range(n)]

for _ in range(n - 1):
    a, b, c = map(int, input().split())
    a, b = a - 1, b - 1
    graph[a].append([b, c])
    graph[b].append([a, c])

q, k = map(int, input().split())
k = k - 1
# kからスタートしてぜんぶたどる
ans_list = [0 for _ in range(n)]


def dfs(node, weight):
    for data in graph[node]:
        if not used[data[0]]:
            used[data[0]] = True
            dfs(data[0], weight + data[1])
    ans_list[node] = weight


used = [False for _ in range(n)]
used[k] = True
dfs(k, 0)

for _ in range(q):
    a, b = map(lambda x: int(x) - 1, input().split())
    print(ans_list[a] + ans_list[b])

File: Practice/atcoder/ABC/070/src/_d.php
<?php
  # pythonで解けない時はphpでやるわ
  # TLEなる.... python以下ですか
  $N = intVal(fgets(STDIN));
  $graph = array_fill(0, $N-1, array());
  for($i=0; $i<$N-1; $i++){
    $abc = array_map("intVal", explode(" ", fgets(STDIN)));
    $a = $abc[0]-1;
    $b = $abc[1]-1;
    $c = $abc[2];
    $graph[$a][] = ["node"=>$b, "weight"=>$c];
    $graph[$b][] = ["node"=>$a, "weight"=>$c];
  }
  # print_r($graph);
  $qk = array_map("intVal", explode(" ", fgets(STDIN)));
  $q = $qk[0];
  $k = $qk[1] - 1;
  $ans_list = array_fill(0, $N-1, 0);


  function dfs($node, $used, $weight){
    global $graph, $N, $ans_list;
    foreach($graph[$node] as $data){
      if(!in_array($data["node"], $used, true)){
        $next_used = $used;
        $next_used[]= $data["node"];
        $i+=1;
        dfs($data["node"], $next_used, $weight + $data["weight"]);
      }
    }
    $ans_list[$node] = $weight;
  }

  $arr = [$k];
  dfs($k, $arr, 0);
  for($i=0; $i<$q; $i++){
    $ab = array_map("intVal", explode(" ", fgets(STDIN)));
    $a = $ab[0] - 1;
    $b = $ab[1] - 1;
    echo $ans_list[$a] + $ans_list[$b] . PHP_EOL;
  }
?>